src/runner.py

```diff
@@ -0,0 +1,576 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manage the lifecycle of runners.
+
+The `Runner` class stores the information on the runners and manages the
+lifecycle of the runners on LXD and GitHub.
+
+The `RunnerManager` class from `runner_manager.py` creates and manages a
+collection of `Runner` instances.
+"""
+
+from __future__ import annotations
+
+import json
+import logging
+import time
+from pathlib import Path
+from typing import Iterable, Optional, Sequence
+
+from errors import LxdError, RunnerCreateError, RunnerError, RunnerFileLoadError, RunnerRemoveError
+from lxd import LxdInstance
+from lxd_type import LxdInstanceConfig
+from runner_type import (
+    GitHubOrg,
+    GitHubRepo,
+    RunnerClients,
+    RunnerConfig,
+    RunnerStatus,
+    VirtualMachineResources,
+)
+from utilities import retry
+
+logger = logging.getLogger(__name__)
+
+
+class Runner:
+    """Single instance of GitHub self-hosted runner.
+
+    Attrs:
+        app_name (str): Name of the charm.
+        path (GitHubPath): Path to GitHub repo or org.
+        proxies (ProxySetting): HTTP proxy setting for juju charm.
+        name (str): Name of the runner instance.
+        exist (bool): Whether the runner instance exists on LXD.
+        online (bool): Whether GitHub marks this runner as online.
+        busy (bool): Whether GitHub marks this runner as busy.
+    """
+
+    runner_application = Path("/opt/github-runner")
```

Comment:
This is a tough one, there are arguments for going both ways:

* If there are no other variables related to this path (e.g., the contents or something like that), adding the `_path` postfix doesn't add a lot of value
* * If there are related variables that are defined, then it can be worthwhile to indicate that one hols the path handle, another the content, another the deserialised content etc

Reply:
minor comment